src/js/components/Calendar/Calendar.js

```diff
@@ -314,61 +325,99 @@ const Calendar = forwardRef(
           }
+          if (activeRangeBoundProp) setActiveRangeBound(activeRangeBoundProp);
         } else {
           // have dates
           const priorDates = dates[0].map(d => new Date(d));
-          const previousDate = new Date(lastSelectedDate || dates[0][0]);
           const selDate = new Date(selectedDate);
           if (selDate.getTime() === priorDates[0].getTime()) {
-            [[, nextDate]] = dates;
-            nextDates = undefined;
+            nextDates = [[undefined, dates[0][1]]];
+            setActiveRangeBound(activeRangeBounds.start);
           } else if (selDate.getTime() === priorDates[1].getTime()) {
-            [[nextDate]] = dates;
-            nextDates = undefined;
-          } else if (selDate.getTime() === previousDate.getTime()) {
-            if (selDate.getTime() < priorDates[0].getTime()) {
+            nextDates = [[dates[0][0], undefined]];
+            setActiveRangeBound(activeRangeBounds.end);
+            if (activeRangeBoundProp) setActiveRangeBound(activeRangeBoundProp);
+          } else if (activeRangeBound === activeRangeBounds.start) {
+            if (selDate.getTime() > priorDates[1].getTime()) {
+              nextDates = [[selectedDate, undefined]];
+            } else {
               nextDates = [[selectedDate, dates[0][1]]];
-            } else if (selDate.getTime() > priorDates[0].getTime()) {
-              nextDates = [[dates[0][0], selectedDate]];
             }
-          } else if (selDate.getTime() < previousDate.getTime()) {
+            setActiveRangeBound(activeRangeBounds.end);
+            if (activeRangeBoundProp) setActiveRangeBound(activeRangeBoundProp);
+          } else if (activeRangeBound === activeRangeBounds.end) {
             if (selDate.getTime() < priorDates[0].getTime()) {
-              nextDates = [[selectedDate, dates[0][1]]];
-            } else if (selDate.getTime() > priorDates[0].getTime()) {
+              nextDates = [[selectedDate, undefined]];
+              setActiveRangeBound(activeRangeBounds.end);
+            } else {
               nextDates = [[dates[0][0], selectedDate]];
+              setActiveRangeBound(activeRangeBounds.start);
             }
-          } else if (selDate.getTime() > previousDate.getTime()) {
-            if (selDate.getTime() > priorDates[1].getTime()) {
-              nextDates = [[dates[0][0], selectedDate]];
-            } else if (selDate.getTime() < priorDates[1].getTime()) {
-              nextDates = [[selectedDate, dates[0][1]]];
-            }
+            if (activeRangeBoundProp) setActiveRangeBound(activeRangeBoundProp);
           }
+          // cleanup
+          if (!nextDates[0][0] && !nextDates[0][1]) nextDates = undefined;
         }
 
         setDates(nextDates);
         if (!dates) setDate(nextDate);
         setActive(new Date(selectedDate));
-        setLastSelectedDate(selectedDate);
-        if (onSelect) onSelect(nextDates || nextDate);
+        let adjustedDates;
+        if (
```

Comment:
Can we put all of this inside the `if (onSelect)`, since `adjustedDates` isn't used at all of `!onSelect`.

Reply:
good call.